### ISSUE TITLE
kubenet/kubelet leaks ips on docker restart

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#34278]
Restart docker and you'll see the number of allocatd ips jumping:
```console
# ls /var/lib/cni/networks/kubenet/ | wc -l
51

# sudo systemctl restart docker.service

# ls /var/lib/cni/networks/kubenet/ | wc -l
69
```
Do it enough times (or with enough pods, do it once) and the node is effectively unschedulable, but reports ready: 
```console
# ls /var/lib/cni/networks/kubenet/ | wc -l
254

# docker ps
CONTAINER ID        IMAGE                                                                  COMMAND                  CREATED             STATUS              PORTS               NAMES
1e60c680bb0f        gcr.io/google_containers/kube-proxy:fa5094377f97111eda3c700c05b313b2   "/bin/sh -c 'kube-pro"   2 minutes ago       Up 2 minutes                            k8s_kube-proxy.84f0299f_kube-proxy-e2e-test-beeps-minion-group-1uut_kube-system_90fd8ba4ce791b234b5cbbba324659b6_a32991a1
68d21830a8d4        gcr.io/google_containers/pause-amd64:3.0                               "/pause"                 2 minutes ago       Up 2 minutes                            k8s_POD.d8dbe16c_kube-proxy-e2e-test-beeps-minion-group-1uut_kube-system_90fd8ba4ce791b234b5cbbba324659b6_a8345dfb
47c4073a177f        bprashanth/kube-dnsmasq-amd64:0.1                                      "/usr/sbin/dnsmasq --"   2 minutes ago       Up 2 minutes                            k8s_dnsmasq.b185c2d1_dnsmasq-e2e-test-beeps-minion-group-1uut_kube-system_66b11a42154e4cd1f207bf98be04a70a_f2725cc2
d92c0f6f80b3        gcr.io/google_containers/pause-amd64:3.0                               "/pause"                 2 minutes ago       Up 2 minutes                            k8s_POD.d8dbe16c_dnsmasq-e2e-test-beeps-minion-group-1uut_kube-system_66b11a42154e4cd1f207bf98be04a70a_858b398a

```

Scheduling pods on it show
```console
Tolerations:	<none>
Events:
  FirstSeen	LastSeen	Count	From						SubObjectPath	Type		Reason		Message
  ---------	--------	-----	----						-------------	--------	------		-------
  5s		1s		4	{kubelet e2e-test-beeps-minion-group-1uut}			Warning		FailedSync	Error syncing pod, skipping: failed to "SetupNetwork" for "echoheaders-3ty5d_default" with SetupNetworkError: "Failed to setup network for pod \"echoheaders-3ty5d_default(997a82ce-8c15-11e6-bbb0-42010af00002)\" using network plugins \"kubenet\": Error adding container to network: no IP addresses available in network: kubenet; Skipping pod"
```

I'm guessing this is because we don't release unused ips, but allocate new ones.
@kubernetes/sig-network @kubernetes/sig-node 

Maybe we can fix this by invoking kubenet everytime kubelet notices a container restart?

**Comment by caseydavenport** [comment]
This also affects CNI - https://github.com/kubernetes/kubernetes/issues/14940

In general, plugins don't get notified when containers are deleted under the kubelet's feet.

**Comment by bprashanth** [comment]
Lets continue discussion on the longer term solution on that bug and keep this for how to surface/handle/cordon/recover such nodes. 

Is there a cheap fix? Maybe just GC backdoor into kubenet?
Can we cordon the node and send an event so users can perform manual recovey?

**Comment by freehan** [comment]
Moving networking logic to runtime thru CRI could make the solution simpler. 

Then the dockershim can keep track of restarts/crashes and invoke CNI delete. 

For `bridge` plugin in particular, delete operation will first release IP. So even if the container is gone, it will release IP and fail later, which is okay for clean up.

**Comment by bprashanth** [comment]
I don't even think we're getting tearDown, probably because something higher up the stack errs out (maybe docker inspect). 

https://github.com/kubernetes/kubernetes/issues/34283 for mitigating impact
https://github.com/kubernetes/kubernetes/issues/34284 for testing

**Comment by freehan** [comment]
```
I don't even think we're getting tearDown, probably because something higher up the stack errs out (maybe docker inspect).
```

kubelet also need to cache the previous container ID in order to do proper tear down.

**Comment by bprashanth** [comment]
well this is where "sync" would shine :)
A periodic sync([]podID) would allow us to purge everything not in that list. 

We can still loop over all the allocated ip/hash buckets, do a docker inspect, and if either the hash doesn't exist, or has an exit code, purge it? The assumption here is that if we've recorded a hash in /var, then it must have existed at one time, and if ps doesn't show the hash now, docker must've restarted. Is there a race? or docker ps perf issues?

**Comment by yifan-gu** [comment]
@freehan Caching the container ID doesn't seem to work well if kubelet restarts after a teardown()fails, but the container is killed.


Some quick ideas here:
1, Do not kill the sandbox (pause container) if teardown() fails, so we can retry later
2, Kill the sandbox without teardown(). Then teardown() when garbage collecting the container, but if teardown() fails, don't remove the container.
3, Try teardown() when killing the sandbox. Try teardown() again before garbage collection, and don't remove the container if teardown() fails.

**Comment by bprashanth** [comment]
I think the base assumption was (erroneously made) that if the kubelet knows about old/new container, the old gets a tearDown just like the new gets a setUp. If the kubelet doesn't know about the old one, all bets are off. If it restarted at the exact wrong moment leaving pods broken, they end up crashing or failed or we just leak, till we have a more complex solution like snapshotting.  

We probably either need tearDown guarantee modulo crash situation (which is starting to sound more complex), a backdoor gc operation or periodic sync.

**Comment by thockin** [comment]
Is this just another case of "if we had a checkpoint we would never lose
state" ?

**Comment by bprashanth** [comment]
No that's the harder issue. 
The more imminent issue is that we don't get a tearDown notification for the old infra container on  container restart because the kubelet doesn't track containerID (it just does a simple name check). Tracking containerID is fickle for the reasons Yifan gave, among others.  

So either: 
* Kubelet gives exact tearDown notifications (tracks containerID, we come up with a better solution for crash recovery)
* Kubelet periodically tells kubenet what is supposed to be running (sync to amend dropped edges)
* Kubenet runs GC for ipam, directly talking to docker ps to figure out what ips are for dead containers (we already record some form of container ID along with each ip, we'd `docker inspect`)

2 feels like the easiest, we should probably do 1 and 2 with a // TODO (https://github.com/kubernetes/kubernetes/issues/489)

**Comment by freehan** [comment]
>well this is where "sync" would shine :)
A periodic sync([]podID) would allow us to purge everything not in that list.

We could trigger a GC sync once encounter a `no IP address` error. That should avoid the race between `docker ps/inspect` and ipam file check. 

Thoughts?

**Comment by bprashanth** [comment]
Can you flesh out the race?
If I find a hash in the file, and docker inspect it: 
* And the container has exited: we have what we want
* Docker says no such container: docker must have restarted, becase at one time it knew the hash

Even if it's just about to reallocate that hash to another container, the new container must get a new ip (so not the one you're deleting) or wait for your delete to get an ip, right? or does kubenet say I got a hash from kubelet, i see a hash on file, I need to return the same ip?

**Comment by freehan** [comment]
>Even if it's just about to reallocate that hash to another container, the new container must get a new ip (so not the one you're deleting) or wait for your delete to get an ip, right?

If we put GC sync inside the mutex protected block, then kubenet will wait till the GC has completed to get a new ip. 

>or does kubenet say I got a hash from kubelet, i see a hash on file, I need to return the same ip?

host-local ipam will allocate a new IP.

**Comment by freehan** [comment]
Okay. I will put together a PR shortly.

**Comment by bprashanth** [comment]
Lets say we don't even lock anything, as a though experiment, what's the race?

**Comment by sampeka** [comment]
I'm also experiencing this problem, is there a workaround for this at the moment? I.e. is there a way to force Kubernetes to release the unused IP addresses?

**Comment by emac08** [comment]
causing major issues for us... is there a resolution/upgrade in the pipeline for gke?

**Comment by philips** [comment]
@freehan Can you explain this a bit more:

> Moving networking logic to runtime thru CRI could make the solution simpler.
> 
> Then the dockershim can keep track of restarts/crashes and invoke CNI delete.

The dockershim is built-in to the kubelet process and then may be a different process in the future. How does this fix the issue?

**Comment by bprashanth** [comment]
Hmm, I think you may be conflating issues. Moving networking logic into the docker-shim is not going to fix the ip leak, the debate is weather the kubelet should even care. This is my understanding of the discussion on the short-term pr (https://github.com/kubernetes/kubernetes/pull/34373).

In today's world, the kubelet lists containers, and if a given container ID (be it name or hash) doesn't show up in the list, it assumes all resources associated with the container don't exist. 

This bug cropped up because restarting docker recycles all IDs.
So say we could checkpoint IDs. The kubelet is now aware of `checkpointStaleID` and has to invoke:
* DockerDeleteAll(checkpointStaleID)
* VolumesDeleteAll(checkpointStaleID)
* CNIDeleteAll(checkpointStaleID)

...

Instead of this, the suggestion was that the kubelet would invoke `RuntimeShimDeleteAll(checkpointStaleID)`, and that would translate to all the mentioned calls in the runtime shim, because only the runtime is really aware of the various bits of state associated with a sandbox. 

Either way, if the runtime says: "this ID doesn't exist", the kubelet assumes all resources associated with the ID don't exist. Even with checkpointing, this would be an issue. 

The short term fix is to reconcile ID<->ip mapping in a GC routine. List all IDs that exist, list all ips, delete the ips that don't match an ID. 

The longer term fix (as I understand it), is the kubelet says "give me a list of IDs", and even if a container is dead but the ip is lying around, the runtime returns that ID, and kubelet invokes `RuntimeShimDeleteAll(ID)`

Such a contract seems pretty simple and makes sense to me, at least at a high level.

**Comment by bprashanth** [comment]
@sampeka @emac08 you can work around it by purging unused ips. 
Enter /var/lib/cni/networks/kubenet on the node and you'll see what's described in https://github.com/kubernetes/kubernetes/issues/34278#issue-181543914. 

You can peek into each file, and match it with a running container:
```console
$ cd  /var/lib/cni/networks/kubenet
$ ls
10.180.2.2  10.180.2.3  10.180.2.4  last_reserved_ip
$ echo foobar > 10.180.2.244
$ for hash in $(tail -n +1 * | grep '^[A-Za-z0-9]*$' | cut -c 1-8); do if [ -z $(docker ps -a | grep $hash | awk '{print $1}') ]; then grep -ir $hash ./; fi; done 
./10.180.2.244:foobar
```
And you can delete them by feeding that list to xargs
```console
$ for hash in $(tail -n +1 * | grep '^[A-Za-z0-9]*$' | cut -c 1-8); do if [ -z $(docker ps -a | grep $hash | awk '{print $1}') ]; then grep -ilr $hash ./; fi; done | xargs rm
```

**Comment by yujuhong** [comment]
> The dockershim is built-in to the kubelet process and then may be a different process in the future. How does this fix the issue?

@philips please see #34974 for the discussion on `PodSandbox` resource cleanup in CRI.

**Comment by miracle2k** [comment]
On my cluster, this is definitely because Docker keeps restarting:

```
-- Logs begin at Mon 2016-10-24 17:29:27 UTC, end at Tue 2016-10-25 14:23:06 UTC. --
Oct 24 17:54:05 gke-production-default-pool-78c646b4-q7ol health-monitor.sh[1426]: Docker daemon failed!
Oct 24 19:23:55 gke-production-default-pool-78c646b4-q7ol health-monitor.sh[1426]: Docker daemon failed!
Oct 24 20:57:06 gke-production-default-pool-78c646b4-q7ol health-monitor.sh[1426]: Docker daemon failed!
```

I can't find the reason why docker is restarting, though.

**Comment by rtoma** [comment]
@bprashanth Thanks for sharing the oneliner to free unused IP addresses! I was able to cure our GKE 1.4.3 cluster from a "no IP addresses available in network" FailedSync loop.

**Comment by dchen1107** [comment]
This is a serious production bug in 1.4 with kubenet, and we should fix and even patch 1.4 IMP. I raised the priority to P0 to reflect the urgency on this one. I also updated another issue at: https://github.com/kubernetes/kubernetes/issues/21656#issuecomment-256142942 in which many users complain this particular bug. 

@freehan has a pending pr https://github.com/kubernetes/kubernetes/pull/34373 to address the issue, but he is on leave for now. @matchstick I am assigning this to you to find another owner. Thanks!

**Comment by bprashanth** [comment]
Yeah the plan was to get that pr in to the next minor releaes. Are you saying this alone qualifies a new release? if not, we have time no?

**Comment by bprashanth** [comment]
The reason time is important is because if the next release is tomorrow, we'd go with the gc hack, but if there's time we can investigate an alternate solution (I haven't dug up if we can in fact just call teardown on old container id for infra container).

**Comment by dchen1107** [comment]
@freehan and I had a brief conversation a while back, and I don't like gc hack. But if we couldn't find another solution in short term or the qualified solution is too complicated for a patch release, I can leave with gc solution in 1.4.

**Comment by bprashanth** [comment]
We're on the same page, I'm holding out in the hope that I'll find a better solution by the time someone announces the 1.4.5 release train. Minhan's pr is pretty much ready modulo nits, so we can get it in an revert too (gross but whatever).

**Comment by yujuhong** [comment]
> The reason time is important is because if the next release is tomorrow, we'd go with the gc hack, but if there's time we can investigate an alternate solution (I haven't dug up if we can in fact just call teardown on old container id for infra container

Tear down the old container id for an infra container should be fine after #35240, and the CRI e2e tests are basically doing that right now. There is no tests for non-kubenet plugins though.

That said, in today's kubelet, the GC hack works for all cases (e.g., infra containers got removed by user directly), and the alternative approach doesn't.

**Comment by bprashanth** [comment]
Oh, so this is essentially fixed in the specific case of docker restart. I'll have to try and repro to confirm. We might still have issues because we can't find the pid of an exited container or something.

**Comment by bprashanth** [comment]
Alright next release is the coming friday, lets just go with the hack.

**Comment by bprashanth** [comment]
the immediate issue is fixed with https://github.com/kubernetes/kubernetes/pull/35816, we have https://github.com/kubernetes/kubernetes/issues/14940 for a better long term fix.

**Comment by edevil** [comment]
What is the current status of this problem? I'm running 1.7.5 with the Kubenet plugin and have run out of IPs because they were not GCed.

I see that the plugin.ipamGarbageCollection() call is still there on the SetUpPod() code but is now guarded by a SupportsLegacyFeatures() condition. This condition is not true in my case since I see no "Starting IP garbage collection" messages in the log.

**Comment by H3rman8** [comment]
Did run into this issue today and refined the workaround due to extraneous linefeed and interface line within the files and containerd instead of docker. Don't know if that's related to my k3os cluster. To test just simply replace 'xargs rm' with 'xargs echo'.

cd /var/lib/cni/networks/cbr0
for hash in $(head -n 1 * | tr -d '\r'| egrep '^[A-Za-z0-9]{64,64}$'); do
if [ -z $(crictl pods --no-trunc | grep $hash | awk '{print $1}') ]; then
grep -ilr $hash ./ | xargs rm
else
echo $(grep -ilr $hash ./) "=>" $(crictl pods --no-trunc | grep $hash|awk '{print $6}')
fi;
done
